File: bench/debug/javascript/01-logic-error/parser.js
const { TOKEN_TYPES } = require('./tokenizer');

const PRECEDENCE = {
  [TOKEN_TYPES.PLUS]: 1,
  [TOKEN_TYPES.MINUS]: 1,
  [TOKEN_TYPES.STAR]: 2,
  [TOKEN_TYPES.SLASH]: 2,
  [TOKEN_TYPES.CARET]: 3,
};

const RIGHT_ASSOCIATIVE = new Set([TOKEN_TYPES.CARET]);

class ASTNode {
  constructor(type, value, left, right) {
    this.type = type;
    this.value = value;
    this.left = left || null;
    this.right = right || null;
  }
}

class Parser {
  constructor(tokens) {
    this.tokens = tokens;
    this.pos = 0;
  }

  current() {
    return this.tokens[this.pos];
  }

  consume(expectedType) {
    const token = this.current();
    if (expectedType && token.type !== expectedType) {
      throw new Error(`Expected ${expectedType} but got ${token.type}`);
    }
    this.pos++;
    return token;
  }

  parseExpression(minPrec = 0) {
    let left = this.parseAtom();

    while (
      this.current().type !== TOKEN_TYPES.EOF &&
      this.current().type !== TOKEN_TYPES.RPAREN &&
      PRECEDENCE[this.current().type] !== undefined &&
      PRECEDENCE[this.current().type] >= minPrec
    ) {
      const op = this.consume();
      const prec = PRECEDENCE[op.type];

      // BUG: For right-associative operators like ^, we should use `prec`
      // (same precedence) so that the recursive call consumes the rest of
      // the chain. Using `prec + 1` forces left-associative grouping.
      // Fix: const nextMinPrec = RIGHT_ASSOCIATIVE.has(op.type) ? prec : prec + 1;
      const nextMinPrec = prec + 1;

      const right = this.parseExpression(nextMinPrec);
      left = new ASTNode('binary', op.value, left, right);
    }

    return left;
  }

  parseAtom() {
    const token = this.current();

    if (token.type === TOKEN_TYPES.NUMBER) {
      this.consume();
      return new ASTNode('number', token.value);
    }

    if (token.type === TOKEN_TYPES.LPAREN) {
      this.consume(TOKEN_TYPES.LPAREN);
      const expr = this.parseExpression(0);
      this.consume(TOKEN_TYPES.RPAREN);
      return expr;
    }

    if (token.type === TOKEN_TYPES.MINUS) {
      this.consume();
      const operand = this.parseAtom();
      return new ASTNode('unary_minus', '-', operand);
    }

    throw new Error(`Unexpected token: ${token.type}`);
  }

  parse() {
    const ast = this.parseExpression(0);
    if (this.current().type !== TOKEN_TYPES.EOF) {
      throw new Error(`Unexpected token after expression: ${this.current().type}`);
    }
    return ast;
  }
}

module.exports = { Parser, ASTNode };

File: bench/debug/cpp/03-crash/parser.cpp
#include "parser.h"
#include <stdexcept>

Parser::Parser(const std::vector<Token>& tokens) : tokens(tokens), pos(0) {}

const Token& Parser::current() const {
    return tokens[pos];
}

const Token& Parser::advance() {
    const Token& tok = tokens[pos];
    if (pos < tokens.size() - 1) pos++;
    return tok;
}

bool Parser::match(TokenType type) {
    if (current().type == type) {
        advance();
        return true;
    }
    return false;
}

// Parse an additive expression: term (('+' | '-') term)*
ASTNode* Parser::parseExpression() {
    ASTNode* left = parseTerm();

    while (current().type == TokenType::Plus ||
           current().type == TokenType::Minus) {
        char op = (current().type == TokenType::Plus) ? '+' : '-';
        advance();
        ASTNode* right = parseTerm();
        left = new ASTNode(op, left, right);
    }

    return left;
}

// Parse a multiplicative expression: primary (('*' | '/') primary)*
//
// BUG: When the left operand is a unary minus node (e.g., from parsing
// "-(3+4)"), this function attempts an "optimization" that extracts the
// inner operand from the unary node, deletes the unary wrapper, and
// negates the inner node's value field. But the inner node is a BinaryOp
// (3+4), not a Number, so setting its value field does nothing useful.
// Worse, deleting the unary wrapper triggers its destructor, which
// recursively deletes its child (the inner BinaryOp). The local pointer
// to the inner node is now dangling. Using it as the left operand of
// the multiplication creates a use-after-free, leading to a crash when
// the evaluator traverses the freed node.
ASTNode* Parser::parseTerm() {
    ASTNode* left = parsePrimary();

    while (current().type == TokenType::Star ||
           current().type == TokenType::Slash) {
        char op = (current().type == TokenType::Star) ? '*' : '/';
        advance();

        // BUG: Attempt to "optimize" unary minus by folding it into
        // the multiplication. Setting value on a BinaryOp node does
        // nothing — the negation is silently lost.
        if (left->type == ASTNode::UnaryOp && left->op == '-') {
            ASTNode* inner = left->left;
            left->left = nullptr;
            delete left;
            inner->value = -inner->value;
            left = inner;
        }

        ASTNode* right = parsePrimary();
        left = new ASTNode(op, left, right);
    }

    return left;
}

// Parse a primary: number, unary minus, or parenthesized expression
ASTNode* Parser::parsePrimary() {
    // Unary minus
    if (current().type == TokenType::Minus) {
        advance();
        ASTNode* operand = parsePrimary();
        return new ASTNode('-', operand);
    }

    // Parenthesized expression
    if (current().type == TokenType::LParen) {
        advance();
        ASTNode* expr = parseExpression();
        if (!match(TokenType::RParen)) {
            throw std::runtime_error("Expected closing parenthesis");
        }
        return expr;
    }

    // Number literal
    if (current().type == TokenType::Number) {
        double val = current().numValue;
        advance();
        return new ASTNode(val);
    }

    throw std::runtime_error(
        "Unexpected token: " + current().text);
}

ASTNode* Parser::parse() {
    ASTNode* result = parseExpression();
    if (current().type != TokenType::End) {
        throw std::runtime_error(
            "Unexpected token after expression: " + current().text);
    }
    return result;
}
